include/deal.II/lac/precondition.h

```diff
@@ -767,61 +816,171 @@ namespace internal
       typename PreconditionerType,
       typename VectorType,
       typename std::enable_if<!has_Tstep<PreconditionerType, VectorType>::value,
                               PreconditionerType>::type * = nullptr>
     void
-    Tstep(const PreconditionerType &preconditioner,
+    Tstep(const MatrixType &        A,
+          const PreconditionerType &preconditioner,
           VectorType &              dst,
-          const VectorType &        src)
+          const VectorType &        src,
+          const double              relaxation,
+          VectorType &              residual,
+          VectorType &              tmp)
     {
-      Assert(false, ExcNotImplemented());
-      (void)preconditioner;
-      (void)dst;
-      (void)src;
+      residual.reinit(dst, true);
+      tmp.reinit(dst, true);
+
+      A.Tvmult(residual, dst);
+      residual.sadd(-1.0, 1.0, src);
+
+      preconditioner.Tvmult(tmp, residual);
+      dst.add(relaxation, tmp);
+    }
+
+    template <typename MatrixType,
+              typename PreconditionerType,
+              typename VectorType>
+    void
+    step_operations(const MatrixType &        A,
+                    const PreconditionerType &preconditioner,
+                    VectorType &              dst,
+                    const VectorType &        src,
+                    const double              relaxation,
+                    VectorType &              tmp1,
+                    VectorType &              tmp2,
+                    const unsigned int        i,
+                    const bool                transposed)
+    {
+      if (i == 0)
+        {
+          if (transposed)
+            preconditioner.Tvmult(dst, src);
+          else
+            preconditioner.vmult(dst, src);
+
+          if (relaxation != 1.0)
+            dst *= relaxation;
+        }
+      else
+        {
+          if (transposed)
+            Tstep(A, preconditioner, dst, src, relaxation, tmp1, tmp2);
+          else
+            step(A, preconditioner, dst, src, relaxation, tmp1, tmp2);
+        }
+    }
+
+    template <typename MatrixType, typename VectorType>
+    void
+    step_operations(const MatrixType &                A,
+                    const DiagonalMatrix<VectorType> &preconditioner,
+                    VectorType &                      dst,
+                    const VectorType &                src,
+                    const double                      relaxation,
+                    VectorType &                      residual,
+                    VectorType &,
+                    const unsigned int i,
+                    const bool         transposed)
+    {
+      if (i == 0)
+        {
+          const auto dst_ptr  = dst.begin();
+          const auto src_ptr  = src.begin();
+          const auto diag_ptr = preconditioner.get_vector().begin();
+
+          for (unsigned int i = 0; i < dst.locally_owned_size(); ++i)
+            dst_ptr[i] = relaxation * src_ptr[i] * diag_ptr[i];
```

Comment:
I guess this won't work for block vectors? But in that case we won't typically use `DiagonalMatrix`, so we should be fine?

Reply:
I could add a check that the vector type is not a block vector? That would be safe?

Reply:
Yes, I think it would be more clear in terms of the error messages as when we have an `auto` variable where we want to access entries by an operator `[]` (or whatever kind of compile/run time error we would get in that case).